### ISSUE TITLE
Python2.7 and Python3 call

### ISSUE THREAD

**Issue by thalesmaoa** [cbm755/octsympy#770]
I'm trying to use  symbolic package in octave with a minimal distro. The problem is that I'm using python3 as default core but octsympy calls for python2.7. Once that happens I got an sympy error since it is installed for python3 and not for python2.7. Is there any form to change the python calling for python3?

**Comment by mtmiller** [comment]
Does setting the `$PYTHON` environment variable to `python3` (in the shell before starting Octave) work?

**Comment by thalesmaoa** [comment]
I wouldn't like to change my systems variables. I need to set it inside
Octave and the reason is simple:

Before I post this issue I've workaround making python link to python3
(/usr/bin/python -> /usr/bin/python3).

It worked just fine, but reflected on my systems packages. Samba_python
needs python2.7 and when I changed the main python binary I got an error.

Another important info is that I'm going to use inside jupyter notebook.
Change local variable inside the shell is not possible.

Any hack suggestions?

Ps: I know I can simple fix it with pip sympy, but that tiny space is
precious.

**Comment by mtmiller** [comment]
Before dismissing it as something you would not like to do, can you verify that setting it does do what it's supposed to?

Then you can ask for help on getting that configuration into Octave and what the proper way to configure it would be.

**Comment by latot** [comment]
Hi, i'm thinking maybe some internal fix?, add some feature to can set the python version, we can just try execute python$VER (actually $VER can be 2 or 3) and check if exist, then list in sympref to can set it.

**Comment by thalesmaoa** [comment]
@mtmiller You are right!

```
$ alias python='/usr/bin/python3'
$ octave-cli
$ pkg load symbolic
$ syms x
```

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named sympy
OctSymPy v2.5.0: this is free software without warranty, see source.
Initializing communication with SymPy using a popen2() pipe.
error: Python cannot import SymPy: have you installed SymPy?
error: called from
    assert_have_python_and_sympy at line 37 column 5
    python_ipc_popen2 at line 78 column 5
    python_ipc_driver at line 58 column 13
    python_cmd at line 164 column 9
    valid_sym_assumptions at line 38 column 10
    assumptions at line 82 column 7
    syms at line 97 column 13
```

```
$ python --version
Python 3.5.2
```

Didn't work!

@latot That would be nice and fix my problem!

**Comment by mtmiller** [comment]
@thalesmaoa an alias is not the same thing as an environment variable, please test the following to be explicit:

    $ export PYTHON=python3
    $ octave-cli
    >> pkg load symbolic
    >> syms x

Either `PYTHON=python3` or `PYTHON=/usr/bin/python3` will work.

If this works correctly, as it should, unlike an alias, then we can help you figure out how to get that setting into your configuration.

**Comment by mtmiller** [comment]
@latot 

> i'm thinking maybe some internal fix?, add some feature to can set the python version

Yes, there could be an enhancement here to look for SymPy under a few different Python interpreters.

> then list in sympref to can set it

What would you add to `sympref`, an option to enable some kind of automatic detection of Python versions? Or an option to select which Python interpreter to use? Because we already deprecated and removed `sympref python` in preference to the environment variable.

**Comment by mtmiller** [comment]
@thalesmaoa I've tested this on my own and I am able to easily switch between Python 2 and Python 3 (both have SymPy on my system) as follows:

    >> syms x
    ## output showing Python 2.7 is loaded
    OctSymPy: Communication established.  SymPy v1.0.
    >> sympref reset
    
    Resetting the communication mechanism
    Closing the Python pipe...
    >> setenv PYTHON python3
    >> syms x
    ## output showing Python 3.5 is loaded
    OctSymPy: Communication established.  SymPy v1.0.
    >> sympref reset
    
    Resetting the communication mechanism
    Closing the Python pipe...
    >> setenv PYTHON python2
    >> syms x
    ## output showing Python 2.7 is loaded
    OctSymPy: Communication established.  SymPy v1.0.

Can you reproduce this? Does this give you enough information on how this is intended to be configured (either in the global environment or within the running Octave session)? Do you still think something needs to be improved?

**Comment by latot** [comment]
mixing this, if the code test python instead?, like first normal python, if not try python 2, if not python3, if we can't call the normal python send a warning.

**Comment by latot** [comment]
and the idea of can set python, if i remember right there we could set python path, but the idea is automatically can detect available python and list them.

**Comment by thalesmaoa** [comment]
> setenv PYTHON python3

Worked! Perfect!
I have no problem to declare this variable before call syms. 

I will try to change .m with this calling.

**Comment by latot** [comment]
personally, set the env manually don't fix this...

**Comment by mtmiller** [comment]
I would close this as resolved, the original question was how to configure octsympy to use Python 3 explicitly, and it works as expected.

@latot if you think it's worth enhancing the automatic Python detection, I think that should be described in a new issue.

**Comment by latot** [comment]
Okis!

**Comment by cbm755** [comment]
thanks everyone, I gather I can close this.
